UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,627 @@
+          GraphicsDev->DeviceId = Fdt32ToCpu (*Data32) & 0xffff;
+        } else if (AsciiStrCmp (TempStr, "subsystem-id") == 0) {
+          Data32                         = (UINT32 *)(PropertyPtr->Data);
+          GraphicsDev->SubsystemVendorId = Fdt32ToCpu (*Data32) >> 16;
+          GraphicsDev->SubsystemId       = Fdt32ToCpu (*Data32) & 0xffff;
+        } else if (AsciiStrCmp (TempStr, "revision-id") == 0) {
+          Data32                  = (UINT32 *)(PropertyPtr->Data);
+          GraphicsDev->RevisionId = (UINT8)Fdt32ToCpu (*Data32);
+        } else if (AsciiStrCmp (TempStr, "bar-index") == 0) {
+          Data32                = (UINT32 *)(PropertyPtr->Data);
+          GraphicsDev->BarIndex = (UINT8)Fdt32ToCpu (*Data32);
+        }
+      }
+    } else if (AsciiStrCmp (NodePtr->Name, "options") == 0) {
+      DEBUG ((DEBUG_INFO, "  Found options node (%08X)", Node));
+
+      for (SubNode = FdtFirstSubnode (Fdt, Node); SubNode >= 0; SubNode = FdtNextSubnode (Fdt, SubNode)) {
+        NodePtr = (FDT_NODE_HEADER *)((CONST CHAR8 *)Fdt + SubNode + Fdt32ToCpu (((FDT_HEADER *)Fdt)->OffsetDtStruct));
+        DEBUG ((DEBUG_INFO, "\n      SubNode(%08X)  %a", SubNode, NodePtr->Name));
+
+        if (AsciiStrnCmp (NodePtr->Name, "upl-images@", AsciiStrLen ("upl-images@")) == 0) {
+          DEBUG ((DEBUG_INFO, "  Found image@ node \n"));
+          //
+          // Build PayloadBase HOB .
+          //
+          PayloadBase = BuildGuidHob (&gUniversalPayloadBaseGuid, sizeof (UNIVERSAL_PAYLOAD_BASE));
+          ASSERT (PayloadBase != NULL);
+          if (PayloadBase == NULL) {
+            return EFI_OUT_OF_RESOURCES;
+          }
+
+          PayloadBase->Header.Revision = UNIVERSAL_PAYLOAD_BASE_REVISION;
+          PayloadBase->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_BASE);
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "addr", &TempLen);
+
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data64       = (UINT64 *)(PropertyPtr->Data);
+            StartAddress = Fdt64ToCpu (*Data64);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  entry"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", StartAddress));
+
+            PayloadBase->Entry = (EFI_PHYSICAL_ADDRESS)StartAddress;
+          }
+        }
+
+        if (AsciiStrnCmp (NodePtr->Name, "upl-params", AsciiStrLen ("upl-params")) == 0) {
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "address_width", &TempLen);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  address_width"));
+            DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+            SizeOfMemorySpace = (UINT8)Fdt32ToCpu (*Data32);
+ #if defined (MDE_CPU_IA32) || defined (MDE_CPU_X64)
+            BuildCpuHob (SizeOfMemorySpace, 16);
+ #else
+            BuildCpuHob (SizeOfMemorySpace, 0);
+ #endif
+          }

```

Comment:
1. We also need to fill BootMode from FDT upl-params->BootMode:
  convert upl-params->boot-mode string to bootmode_enum
  (EFI_HOB_HANDOFF_INFO_TABLE) (UINTN) mHobList->BootMode = bootmode_enum;

2. later UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES->ResourceAssigned HOB data should be filled by upl-params->pci-enum-done (when present, set to 1)